### DISCUSSION TITLE
Cache flags

### DISCUSSION THREAD

**Discussion by KyleAMathews** [gatsbyjs/gatsby#28331] [Umbrella Discussions]
## Cache clearing

We're investigating ways to be smarter about clearing caches. [Initial PR](https://github.com/gatsbyjs/gatsby/pull/28334)

Currently we delete the entire cache when your gatsby-node.js, gatsby-config.js, or package versions change or you run gatsby clean. This is unnecessarily aggressive — changing these things don't invalidate most caches. The current cache clearing behavior means very expensive operations like downloading and processing images have to be repeated quite often which makes iterating on the website a lot more frustrating than it should.

There are currently two cache clearing flags you can try:

- `PRESERVE_FILE_DOWNLOAD_CACHE` prevents file downloads from being deleted during cache clearing events (other than `gatsby clean` which still deletes the entire cache).
- `DEV_WEBPACK_CACHE` prevents the webpack cache from deleted during `gatsby develop` with the same caveat for `gatsby clean`

We will continue to clear other plugin caches e.g. those created by source & transformer plugins. We're looking at ways to safely making more selective clearing those caches as well.

To enable the new cache behavior, add to your gatsby-config.js:

```js
module.exports = {
  flags: {
    PRESERVE_FILE_DOWNLOAD_CACHE: true,
    DEV_WEBPACK_CACHE: true
  },
  plugins: [...]
}
```

These will also be enabled if you add the `FAST_DEV` flag which enables a number of WIP ways to speed up the dev server.

## Utilising webpack persistent caching

In addition to preserving webpack caches when invalidating gatsby cache, `PRESERVE_FILE_DOWNLOAD_CACHE` flag (since `gatsby@3.4.0`) also utilize [webpack persitent caching](https://webpack.js.org/blog/2020-10-10-webpack-5-release/#persistent-caching) for production builds. This makes webpack steps in subsequent `gatsby build` significantly faster.

**Comment by aaronadamsCA** [comment]
FYI, I'm enabling `FAST_DEV` on Gatsby 2.28.0 and the reporter tells me these two flags are not enabled.

```js
// gatsby-config.js
flags: {
  FAST_DEV: true,
  FAST_REFRESH: true,
},
```

```txt
info The following flags are active:
- FAST_DEV · Enable all experiments aimed at improving develop server start time
- FAST_REFRESH · (Umbrella Issue (​https://github.com/gatsbyjs/gatsby/discussions/28390​)) · Use React Fast Refresh instead of the legacy react-hot-loader for instantaneous feedback in your development server. Recommended for versions of React >= 17.0.
- DEV_SSR · (Umbrella Issue (​https://github.com/gatsbyjs/gatsby/discussions/28138​)) · SSR pages on full reloads during develop. Helps you detect SSR bugs and fix them without needing to do full builds.
- QUERY_ON_DEMAND · (Umbrella Issue (​https://github.com/gatsbyjs/gatsby/discussions/27620​)) · Only run queries when needed instead of running all queries upfront. Speeds starting the develop server.
- LAZY_IMAGES · EXPERIMENTAL · (Umbrella Issue (​https://github.com/gatsbyjs/gatsby/discussions/27603​)) · Don't process images during development until they're requested from the browser. Speeds starting the develop server.

There are 2 other flags available that you might be interested in:
- PRESERVE_WEBPACK_CACHE · (Umbrella Issue (​https://github.com/gatsbyjs/gatsby/discussions/28331​)) · Don't delete webpack's cache when changing gatsby-node.js & gatsby-config.js files.
- PRESERVE_FILE_DOWNLOAD_CACHE · (Umbrella Issue (​https://github.com/gatsbyjs/gatsby/discussions/28331​)) · Don't delete the downloaded files cache when changing gatsby-node.js & gatsby-config.js files.
```

**Reply by KyleAMathews** [reply]
oops, we didn't actually add it yet to FAST_DEV. Thanks for the reminder! PR https://github.com/gatsbyjs/gatsby/pull/28447

**Comment by pelikhan** [comment]
Nice call out in the CLI, it's a great way to discover those goodies. The instructions are not specific enough though. Not sure where to put flags.
```
To do so, add to your gatsby-config.js:

flags: {
  PRESERVE_WEBPACK_CACHE: true,
}
```

**Reply by KyleAMathews** [reply]
Is the instructions in the body of the discussion more clear? Thanks for letting us know — we'll keep iterating on the instructions to make them more clear.

**Reply by pelikhan** [reply]
This was much clearer:
```
module.exports = {
  flags: {
    PRESERVE_FILE_DOWNLOAD_CACHE: true,
    PRESERVE_WEBPACK_CACHE: true
  },
  plugins: [...]
}
```

**Comment by krismorf** [comment]


**Reply by KyleAMathews** [reply]
Can you add a screenshot of the error?

**Reply by krismorf** [reply]
Apologies, please delete!

**Comment by jonnyeom** [comment]
Im experiencing an issue where Build / Develop cli logs are not visible if one of the flags are set.

## Description

When one of the new dev flags are set,
During Gatsby develop / build, the console output is hidden / not shown.
![image](https://user-images.githubusercontent.com/19575068/104930281-f1a1f900-5972-11eb-8e65-2917c29d1a44.png)


### Steps to reproduce

On the latest cli code, in `gatsby-config.js`, set one of the new fast_dev flags.
```
  flags: { PRESERVE_WEBPACK_CACHE: true },
  plugins: [
    {...
```


### Expected result

Notice about flags are shown,
An then regular build logs are displayed.


### Actual result
Notice about flags are shown,
Regular build logs are displayed but the text is empty / not visible.

Anyone else?

**Reply by Hahlh** [reply]
Same for me. I updated and the behavior started. Unfortunatelty I can seem to fix it temporarily by setting the flags to false or did I miss any?
![image](https://user-images.githubusercontent.com/45669613/107418070-8503bf80-6b16-11eb-9a49-4549917bf4fb.png)
![image](https://user-images.githubusercontent.com/45669613/107418085-892fdd00-6b16-11eb-89e7-89d57bb70a0f.png)

Edit: Added PARALLEL_SOURCING to the mix without success on gatsby-cli@latest.

**Reply by Hahlh** [reply]
Downgrading gatsby-cli to 2.15.1 resolved the issue, testing 2.16.1 reintroduced it.

**Reply by ethannkschneider** [reply]
I have also experienced this issue on Gatsby v 2.31.0

**Comment by nrandell** [comment]
I've been having some issues on windows where it doesn't seem to preserve the download cache. I think it's due to how 'del' works. I've added in a line to not delete the caches folder and it appears to work.

`deleteGlobs.push(!${cacheDirectory}/caches)` (line 419 of `services/initialise.ts`


 I think this is sort of indicated by the documentation for del https://github.com/sindresorhus/del#beware


I believe it is working for me because my builds are now faster if I have this change in my locally modified node_modules.

**Reply by nrandell** [reply]
Looking into this a bit more, I found that wasn't quite right. This change in initialise.ts seems to work better.


          // Add gatsby-source-filesystem
          deleteGlobs.push(`!${cacheDirectory}/caches`);
          deleteGlobs.push(`${cacheDirectory}/caches/*`);
          deleteGlobs.push(`!${cacheDirectory}/caches/gatsby-source-filesystem`);

This may only be a windows thing though!

**Reply by KyleAMathews** [reply]
I tested the negations on my mac and it seems to work fine. Could you try this on your site? It lists what files would be deleted by the delete globs in initialize.js.

```js
const del = require(`del`)

;(async () => {
  const result = await del(
    // Replace the paths with what your test site would generate.
    [
      "/Users/kylemathews/programs/gatsby/benchmarks/huge-nodes/.cache/**",
      "/Users/kylemathews/programs/gatsby/benchmarks/huge-nodes/.cache/*/",
      "!/Users/kylemathews/programs/gatsby/benchmarks/huge-nodes/.cache/webpack",
      "!/Users/kylemathews/programs/gatsby/benchmarks/huge-nodes/.cache/caches/gatsby-source-filesystem",
    ],
    { dryRun: true }
  )

  console.log(result.filter(r => r.includes(`webpack`)))
  console.log(result.filter(r => r.includes(`filesystem`)))
})()
```

**Reply by nrandell** [reply]
Thanks for this. I just tried your version and it showed the following output
```
[
  'D:\\source\\bond\\bond-london-portfolio-site\\.cache\\caches\\webpack-theme-component-shadowing'
]
[]
```

I then added another log message in

```
  console.log(result.filter((r) => r.includes(`caches`)));
```

And got these results
```
[
  'D:\\source\\bond\\bond-london-portfolio-site\\.cache\\caches\\webpack-theme-component-shadowing'
]
[]
[
  'D:\\source\\bond\\bond-london-portfolio-site\\.cache\\caches',
  'D:\\source\\bond\\bond-london-portfolio-site\\.cache\\caches\\bond-gatsby-source-graphcms',
  'D:\\source\\bond\\bond-london-portfolio-site\\.cache\\caches\\bundle-optimisations',
  'D:\\source\\bond\\bond-london-portfolio-site\\.cache\\caches\\default-site-plugin',
  'D:\\source\\bond\\bond-london-portfolio-site\\.cache\\caches\\dev-404-page',
  'D:\\source\\bond\\bond-london-portfolio-site\\.cache\\caches\\gatsby-plugin-gatsby-cloud',
  'D:\\source\\bond\\bond-london-portfolio-site\\.cache\\caches\\gatsby-plugin-image',
  'D:\\source\\bond\\bond-london-portfolio-site\\.cache\\caches\\gatsby-plugin-manifest',
  'D:\\source\\bond\\bond-london-portfolio-site\\.cache\\caches\\gatsby-plugin-page-creator',
  'D:\\source\\bond\\bond-london-portfolio-site\\.cache\\caches\\gatsby-plugin-postcss',
  'D:\\source\\bond\\bond-london-portfolio-site\\.cache\\caches\\gatsby-plugin-sharp',
  'D:\\source\\bond\\bond-london-portfolio-site\\.cache\\caches\\gatsby-plugin-typescript',
  'D:\\source\\bond\\bond-london-portfolio-site\\.cache\\caches\\gatsby-transformer-sharp',
  'D:\\source\\bond\\bond-london-portfolio-site\\.cache\\caches\\generate-typings',
  'D:\\source\\bond\\bond-london-portfolio-site\\.cache\\caches\\internal-data-bridge',
  'D:\\source\\bond\\bond-london-portfolio-site\\.cache\\caches\\load-babel-config',
  'D:\\source\\bond\\bond-london-portfolio-site\\.cache\\caches\\prod-404',
  'D:\\source\\bond\\bond-london-portfolio-site\\.cache\\caches\\webpack-theme-component-shadowing'
]
```

Note the first entry deletes the `caches' folder!

If I add back in to not delete the `caches` folder, but delete everything under apart from the `gatsby-source-filesystem folder`.

```
   "!D:/source/bond/bond-london-portfolio-site/.cache/caches",
   "D:/source/bond/bond-london-portfolio-site/.cache/caches/*",
```

Which results in

```
[
  'D:\\source\\bond\\bond-london-portfolio-site\\.cache\\caches\\webpack-theme-component-shadowing'
]
[]
[
  'D:\\source\\bond\\bond-london-portfolio-site\\.cache\\caches\\bond-gatsby-source-graphcms',
  'D:\\source\\bond\\bond-london-portfolio-site\\.cache\\caches\\bundle-optimisations',
  'D:\\source\\bond\\bond-london-portfolio-site\\.cache\\caches\\default-site-plugin',
  'D:\\source\\bond\\bond-london-portfolio-site\\.cache\\caches\\dev-404-page',
  'D:\\source\\bond\\bond-london-portfolio-site\\.cache\\caches\\gatsby-plugin-gatsby-cloud',
  'D:\\source\\bond\\bond-london-portfolio-site\\.cache\\caches\\gatsby-plugin-image',
  'D:\\source\\bond\\bond-london-portfolio-site\\.cache\\caches\\gatsby-plugin-manifest',
  'D:\\source\\bond\\bond-london-portfolio-site\\.cache\\caches\\gatsby-plugin-page-creator',
  'D:\\source\\bond\\bond-london-portfolio-site\\.cache\\caches\\gatsby-plugin-postcss',
  'D:\\source\\bond\\bond-london-portfolio-site\\.cache\\caches\\gatsby-plugin-sharp',
  'D:\\source\\bond\\bond-london-portfolio-site\\.cache\\caches\\gatsby-plugin-typescript',
  'D:\\source\\bond\\bond-london-portfolio-site\\.cache\\caches\\gatsby-transformer-sharp',
  'D:\\source\\bond\\bond-london-portfolio-site\\.cache\\caches\\generate-typings',
  'D:\\source\\bond\\bond-london-portfolio-site\\.cache\\caches\\internal-data-bridge',
  'D:\\source\\bond\\bond-london-portfolio-site\\.cache\\caches\\load-babel-config',
  'D:\\source\\bond\\bond-london-portfolio-site\\.cache\\caches\\prod-404',
  'D:\\source\\bond\\bond-london-portfolio-site\\.cache\\caches\\webpack-theme-component-shadowing'
]
```

Note the cache folder is not deleted.

It would be interesting to see if the mac deletes the caches folder.

**Reply by KyleAMathews** [reply]
ooo I see! Yeah, macs have the same behavior as the `.cache/caches` folder is deleted as well without your change. Sweet, want to PR if your fix so we can get this sorted?

**Reply by KyleAMathews** [reply]
Sucks this has been broken all this time 😬

**Reply by KyleAMathews** [reply]
But thanks for investigating and finding the problem/fix

**Reply by nrandell** [reply]
No problems - PR https://github.com/gatsbyjs/gatsby/pull/31069 created

**Reply by KyleAMathews** [reply]
So I was testing our your PR against an image from unsplash and it still wasn't working. I looked into it more and realized that we've only ever added support for `etag` caching — one of multiple options a server might use. So even with this, there'll be plenty of images which won't get cached correctly.

Looking around, I think we should switch to using https://www.npmjs.com/package/make-fetch-happen as it promises us — "Built-in request caching following full HTTP caching rules (Cache-Control, ETag, 304s, cache fallback on error, etc)."

**Reply by KyleAMathews** [reply]
heh actually `got` has caching support we can use so need to switch — https://github.com/sindresorhus/got#cache-1

I tested it quickly against the unsplash URLs and it works as expected

**Reply by KyleAMathews** [reply]
I'll work on a PR to add caching support next week

**Reply by KyleAMathews** [reply]
hmmm actually (another yet again) — https://github.com/kornelski/http-cache-semantics is what we want. The "Got" cache means that we'd need to read & write to our cache each time — `http-cache-semantics` is lower-level so would let us simply look at the previous response headers and often just say we need to do nothing (e.g. the server responded with `cache-control: max-age=604800`).

So this would work like our current system but have much more robust support for different caching setups. I might just post an issue though as I think the `etag` works with our main integrations e.g. Drupal/WordPress/Contentful/etc as all these set `etag`

**Reply by KyleAMathews** [reply]
One plus to changing is right now we do revalidate every time where w/ the above, we often wouldn't — which would save a lot of http requests.

**Comment by klittlepage** [comment]
## Description

When using `gatsby-plugin-typescript` in a yarn2 project with `PRESERVE_WEBPACK_CACHE=true` errors of the form:

```text
w> [webpack.cache.PackFileCacheStrategy/webpack.FileSystemInfo] Resolving 'gatsby-plugin-typescript/gatsby-node' in /tmp/repo/node_modules/gatsby/dist/utils for build dependencies doesn't lead to expected result '/tmp/repo/node_modules/gatsby-plugin-typescript/gatsby-node.js', but to '/tmp/repo/node_modules/gatsby/node_modules/gatsby-plugin-typescript/gatsby-node.js' instead. Resolving dependencies are ignored for this path.
<w>  at unknown 4 gatsby-plugin-typescript/gatsby-node
<w>  at file dependencies /tmp/repo/node_modules/gatsby/dist/utils/api-runner-node.js
<w>  at file /tmp/repo/node_modules/gatsby/dist/utils/api-runner-node.js
<w>  at file dependencies /tmp/repo/node_modules/gatsby/dist/utils/webpack.config.js
<w>  at file /tmp/repo/node_modules/gatsby/dist/utils/webpack.config.js
<w>  at resolve commonjs /tmp/repo/node_modules/gatsby/dist/utils/webpack.config.js
```

are emitted at build time. 

### Steps to reproduce

A minimal reproduction is available [here](https://github.com/klittlepage/gatsby-yarn2-webpack-issue)

Running `yarn build` will result in the error above.

### Expected result

Successful build-time module resolution.

### Actual result

An error at build time.

### Environment

```js
const flags = {
  PRESERVE_WEBPACK_CACHE: true
};
```

```
System:
    OS: macOS 11.3
    CPU: (16) x64 Intel(R) Core(TM) i9-9880H CPU @ 2.30GHz
    Shell: 5.8 - /bin/zsh
  Binaries:
    Node: 14.17.0 - /usr/local/opt/node@14/bin/node
    Yarn: 2.4.1 - /usr/local/bin/yarn
    npm: 6.14.13 - /usr/local/opt/node@14/bin/npm
  Languages:
    Python: 3.8.5 - /Users/klittlepage/opt/miniconda3/bin/python
  Browsers:
    Chrome: 90.0.4430.212
    Firefox: 88.0.1
    Safari: 14.1
  npmPackages:
    gatsby: ^3.6.1 => 3.6.1
    gatsby-plugin-typescript: ^3.6.0 => 3.6.0
```

**Comment by chuckdries** [comment]
I got the "Hi from the Gatsby maintainers!" message about the PRESERVE_WEBPACK_CACHE flag in a CI build. Should I check the .cache folder into source control to take advantage of this? It's in my .gitignore but I'm not sure if I put it there or Gatsby did.

**Reply by KyleAMathews** [reply]
Don't check it in — just setup your CI to preserve and restore the `.cache` and `public` folders in between builds.

**Comment by phil-lgr** [comment]
Can we have support for gatsby config files in `.ts` and `.tsx` format? 

I am using gatsby with ts-node and I get this error:

```txt
Error: <w> [webpack.cache.PackFileCacheStrategy] Caching failed for pack: Error: Can't resolve '/Users/phil/Code/newrade-core/packages/ze-design-website/gatsby-node.js'
```

Also I tried to disable or modify the `.cache` webpack config in `onCreateWebpackConfig` but as of `v3.13.0` it does not work anymore

Relevant code is here: https://github.com/gatsbyjs/gatsby/blob/dac2b73ec57077a4a74baa652eff6b3cabf1c1d3/packages/gatsby/src/utils/webpack.config.js#L863

**Reply by phil-lgr** [reply]
My current cache workaround:

```ts
  config.cache = {
    type: 'filesystem',
    buildDependencies: {
      config: [
        __filename,
        path.resolve(process.cwd(), 'package.json'),
        path.resolve(process.cwd(), 'tsconfig.json'),
        ...store
          .getState()
          .flattenedPlugins.filter((plugin: any) =>
            plugin.nodeAPIs.includes(`onCreateWebpackConfig`)
          )
          .map((plugin: any) => {
            const pluginPath = path.join(plugin.resolve, `gatsby-node.js`);
            const pluginPathTs = path.join(plugin.resolve, `gatsby-node.ts`);
            let foundPluginPath;

            try {
              fs.readFileSync(pluginPath);
              foundPluginPath = pluginPath;
              return foundPluginPath;
            } catch (error: any) {
              reporter.info(`[${pluginOptions.pluginName}] could not find file: ${pluginPath}`);
            }

            try {
              reporter.info(`[${pluginOptions.pluginName}] trying file in ts: ${pluginPathTs}`);
              fs.readFileSync(pluginPathTs);
              reporter.info(`[${pluginOptions.pluginName}] found file: ${pluginPathTs}`);
              foundPluginPath = pluginPathTs;
            } catch (error: any) {
              reporter.info(`[${pluginOptions.pluginName}] could not find file: ${pluginPathTs}`);
            }
            return foundPluginPath;
          }),
      ],
    },
  };
```

Output:

```
info [gastby-plugin-core] could not find file: /Users/phil/Code/newrade-core/packages/core-gatsby-config/src/plugins/gatsby-plugin-core/gatsby-node.js
info [gastby-plugin-core] trying file in ts: /Users/phil/Code/newrade-core/packages/core-gatsby-config/src/plugins/gatsby-plugin-core/gatsby-node.ts
info [gastby-plugin-core] found file: /Users/phil/Code/newrade-core/packages/core-gatsby-config/src/plugins/gatsby-plugin-core/gatsby-node.ts
info [gastby-plugin-core] could not find file: /Users/phil/Code/newrade-core/packages/core-gatsby-config/src/plugins/gatsby-plugin-treat/gatsby-node.js
info [gastby-plugin-core] trying file in ts: /Users/phil/Code/newrade-core/packages/core-gatsby-config/src/plugins/gatsby-plugin-treat/gatsby-node.ts
info [gastby-plugin-core] found file: /Users/phil/Code/newrade-core/packages/core-gatsby-config/src/plugins/gatsby-plugin-treat/gatsby-node.ts
info [gastby-plugin-core] could not find file: /Users/phil/Code/newrade-core/packages/core-gatsby-config/src/plugins/gatsby-plugin-vanilla/gatsby-node.js
info [gastby-plugin-core] trying file in ts: /Users/phil/Code/newrade-core/packages/core-gatsby-config/src/plugins/gatsby-plugin-vanilla/gatsby-node.ts
info [gastby-plugin-core] found file: /Users/phil/Code/newrade-core/packages/core-gatsby-config/src/plugins/gatsby-plugin-vanilla/gatsby-node.ts
```

It replaces the cache entries for gatsby config files with the correct .ts versions

**Reply by phil-lgr** [reply]
The above workaround works for v3.13.0, just make sure that you set the `.cache` prop and `actions.replaceWebpackConfig(config);` for every stage

**Comment by FraserThompson** [comment]
Been a while since something was added to this discussion... But I still think there's a major issue with overly aggressive clearing of thumbnail cache for thumbnails generated by `gatsby-plugin-sharp`. Ideally the only thing which would clear the thumbnail cache would be changes to sharp settings or modules, if I add a frontend dependency to package.json there's no reason why it should have to re-generate all my thumbnails (which in my case is a "leave it overnight" type ordeal).

This was discussed here https://github.com/gatsbyjs/gatsby/issues/24822 but then the issue was closed because part of the problem was addressed by Query on Demand and Lazy Image flags - but this only addresses the issue of slow dev server builds, the issue of production builds was not addressed.

Ideally we could have something like PRESERVE_FILE_DOWNLOAD_CACHE but for image thumbnails.

**Reply by violy** [reply]
Yes, it's an idea… I've also lost sooooooo… many time ! 
May be with a principle of cache container (with a name) to clear specific parts / usages of the cache.
